### PR TITLE
Add a nix build and update lean code

### PR DESCRIPTION
**Pull request by Anderssorby**
This adds a nix build and makes the code run under current lean4.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    lean = {
+      url = github:leanprover/lean4;
+    };
+    nixpkgs.url = github:nixos/nixpkgs/nixos-21.05;
+    utils = {
+      url = github:yatima-inc/nix-utils;
+      inputs.nixpkgs.follows = "nixpkgs";
+    };
+    # A lean dependency
+    lean-ipld = {
```
Comment (gebner) [inline flake.nix]: This dependency is not necessary here, right?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Olean format parsing in Lean
+
+Build with `nix build .`
```
Comment (gebner) [inline README.md]:
```suggestion
Build with `lake build` or `nix build .`
```

Please also update the `lean-toolchain` file.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,74 @@
+{
+  description = "Olean binary format parser in Lean";
+
+  inputs = {
+    lean = {
+      url = github:leanprover/lean4;
```
Comment (gebner) [inline flake.nix]: Can you pin this to the same nightly as the `lean-toolchain` file?

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      # Compile dependencies with the same lean version
+      inputs.lean.follows = "lean";
+    };
+  };
+
+  outputs = { self, lean, utils, nixpkgs, lean-ipld }:
+    let
+      supportedSystems = [
+        # "aarch64-linux"
+        # "aarch64-darwin"
```
Comment (gebner) [inline flake.nix]: These two should work.

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        "i686-linux"
```
Comment (gebner) [inline flake.nix:20]: This one most likely doesn't.

**Review by gebner — COMMENTED** [review]
I have to say the nix setup requires a *lot* of boilerplate (my nixos config flake is shorter...).  I also can't promise to keep the flake up-to-date and working.

File: flake.nix (hunk 3 above)
Comment (Anderssorby) [reply flake.nix]: I couldn't set it specifically since nix uses git and there are no nightly tags. So I sat it to the latest commit.

**Comment by Anderssorby** [comment]
@gebner I can keep the flake up to date as needed. The verbosity of nix is paid back by it's reproducibility. I'm also working on a Lean implementation of nix in Nix.lean.

File: Oleanparser/Object.lean [hunk 6]
```diff
@@ -75,4 +75,4 @@ unsafe def reprCore : Obj → ReprM Format
     return newDeclId
 
 unsafe instance : Repr Obj where
-  reprPrec o := Repr.addAppParen o.reprCore.run
+  reprPrec o := Repr.addAppParen "Obj _"
```
Comment (gebner) [inline Oleanparser/Object.lean]: Please revert this accidental change here.  The `Demo.lean` file contains an end-to-end demo, the last `#eval` there should print the contents of `testData`.

Comment (Anderssorby) [reply Oleanparser/Object.lean]: I had to fix OfNat Any, but now it works. Hope to be of assistance in this project.

Comment (gebner) [reply Oleanparser/Object.lean]: Thanks!

Comment (gebner) [reply Oleanparser/Object.lean]: Oops, apparently I was a bit too quick with merging.  The OfNat instance you added wasn't quite what I expected (it converted every natural number to `()`).  I've changed it `⟨_, n⟩` which was the previous behavior.

**Pull request merged**